content/reference/dos-mitigation.md

```diff
@@ -0,0 +1,197 @@
+---
+title: "DOS Mitigation"
+weight: 4
+---
+
+In general we have to design our protocols to be resilient to malicious or
+misbehaving nodes. libp2p offers reasonable defaults and primitives tools let the
+end application developer design resilient peer-to-peer programs.
+
+# General tactics that work across libp2p implementations
+
+## Limit the number of concurrent streams your protocol needs
+
+Each stream has some resource cost associated with it. Depending on the
+transport and multiplexer, this can be bigger or smaller. Try to avoid having
+too many concurrent streams open per peer for your protocol. Instead try to
+limit the maximum number of concurrent streams to something reasonable (surely
+you don't need >512 streams open at once for a peer?). Multiple concurrent
+streams can be useful for logic or to avoid [Head-of-line
+blocking](https://en.wikipedia.org/wiki/Head-of-line_blocking).
+
+Using a stream for a short period of time and then closing it is fine. It's
+really the number of _concurrent_ streams that you need to be careful of.
+
+## Reduce blast radius
+
+If you can split up your libp2p application into multiple separate processes you
+can increase the resiliency of your overall system. For example your node may
+have to help achieve consensus and respond to user queries. By splitting this up
+into two processes you now rely on the OS’s guarantee that the user query
+process won’t take down the consensus process.
+
+## Fail2ban
+
+If you can log when a peer is misbehaving or malicious, you can then hook up
+those logs to fail2ban and have fail2ban manage your firewall to automatically
+block misbehaving nodes. Go libp2p includes some builtin support for this
+usecase. More details below.
+
+
+# Tactics for Go libp2p
+
+This section is meant for an operator of a go-libp2p application. In the future
+we hope that this advice will work equally well amongst all implementations.
+
+## Resource Manager
+
+go-libp2p includes a powerful resource manager that keeps track of every active
+connection and stream. It exposes some useful metrics that can help you tune
+your resource limits. It’s important to keep an eye on the number of blocked
+resource requests, if you are constantly blocking resources during normal
+operation your limits are too low. But a very rare spike is a good sign that the
+resource manager protected your node from something. You can hook these metrics
+up to Prometheus (or any other opencensus exporter) and Grafana (we even include
+a prebuilt dashboard). See
+[https://github.com/libp2p/go-libp2p-resource-manager/tree/master/obs/grafana-dashboards](https://github.com/libp2p/go-libp2p-resource-manager/tree/master/obs/grafana-dashboards)
+for more details on how to set this up.
+
+![blocked-resource-screenshot](./assets/blocked-resource-screenshot.png)
+
+## Rate limit incoming connections
+
+Depending on your use case, it can help to limit the number of inbound
+connections. You can use go-libp2p's
```

Comment:
Do we give guidance on when to use this mechanism vs. the resource manager?  I see this is a good hook for custom logic, but it seems like what Prysm is doing could be covered by go-libp2p resource manager right?

Reply:
> Do we give guidance on when to use this mechanism vs. the resource manager?

No, but I could add something here.

> I see this is a good hook for custom logic, but it seems like what Prysm is doing could be covered by go-libp2p resource manager right?

Not really. If you're trying to avoid an adversary that can connect to you and give you a ton of work to do all at once the rcmgr doesn't protect at all. This attack can easily be mitigated by rate limiting though.

Not all applications will want this rate limiting, or they may want to rate limit certain things (e.g. something in the protocol rather than in the connections). For example, if I'm Google I wouldn't want to rate limit any new connection to me. I would rather rate limit work per connection.

Should the rcmgr do this? I don't think so. It's not directly related to limiting the resources being used, and if it can be handled by a smaller component that already exists the better.

I hope that makes sense, but happy to expand more as well.